### ISSUE TITLE
algorithm detection

### ISSUE THREAD

**Issue by jeacott1** [auth0/auth0-spring-security-api#54]
Hi,
 is there any reason the line below doesn't just check the alg header to discover the proper encoding instead of only being able to configure this for a single type of jwt?
its tedious to have to configure alternate beans for RSA and HMAC support

 https://github.com/auth0/auth0-spring-security-api/blob/7818c506cd5ce88a6f6345e199a67553069595d0/lib/src/main/java/com/auth0/spring/security/api/JwtAuthenticationProvider.java#L64

**Comment by jeacott1** [comment]
thanks for the link! and for the general case it makes sense to me, but I'm still not sure I understand why it would be bad to modify that line to something like:
 if (secret != null && algorithmIsHMAC(alg)) {...

it is more restrictive than current, and would enable the use of both hmac and rsa tokens making integration testing less painful with the same configuration.

**Comment by lbalmaceda** [comment]
In any case, HMAC works with a secret in comparison to RSA which accepts only public and private keys that's why the method's signature differs, so if `secret != null` you're 100% in a HMAC scenario. This is also [called internally](https://github.com/auth0/auth0-spring-security-api/blob/059563c556218779d7eec99c9b3c8f9f26a3d5ec/lib/src/main/java/com/auth0/spring/security/api/JwtWebSecurityConfigurer.java#L74) from a helper class (which is how we propose this lib to be used) so in an ideal scenario you're not interacting directly with that class.

**Comment by jeacott1** [comment]
I acknowledge that HMAC requires a secret, and as a result of that its easy to use for integration testing.
If that line filtered on the hmac alg as well as the secret existing I could have my cake and eat it!
hmac would work, rsa would work, and no bunnies will have died in the process. I wont need to filter on profiles or env vars etc to select the correct config for the environment and a bunch of developer overhead dealing with special cases will disappear.
Do you think we can get this alteration in?

**Comment by lbalmaceda** [comment]
There's no "alg" value to switch on. Like I said, the class only supports HS256 and RS256. Each is accessible using the [byte array](https://github.com/auth0/auth0-spring-security-api/blob/7818c506cd5ce88a6f6345e199a67553069595d0/lib/src/main/java/com/auth0/spring/security/api/JwtAuthenticationProvider.java#L28) or the [JwkProvider](https://github.com/auth0/auth0-spring-security-api/blob/7818c506cd5ce88a6f6345e199a67553069595d0/lib/src/main/java/com/auth0/spring/security/api/JwtAuthenticationProvider.java#L35) constructor respectively. That's the way we know which algorithm to use later. So again, if the verification requires to use HS256 the secret would be not null as in the [line](https://github.com/auth0/auth0-spring-security-api/blob/7818c506cd5ce88a6f6345e199a67553069595d0/lib/src/main/java/com/auth0/spring/security/api/JwtAuthenticationProvider.java#L64) you originally shared.

Since the class is not going to add support for any other algorithm, since those are the ones used by Auth0, you can rest assured that the first IF clause:

```java
if (secret != null) { 
}
```

is the same as / would mean the same as

```java
if (secret != null && algorithmIsHMAC(alg)) {
}
```

**Comment by jeacott1** [comment]
"is the same as / would mean the same as

if (secret != null && algorithmIsHMAC(alg)) {
}"

is not true. with the addition of algorithmIsHMAC(alg) 
I can set a secret AND rsa tokens will still work properly. this is what I want.

**Comment by lbalmaceda** [comment]
> I can set a secret AND rsa tokens will still work properly

That's not open for discussion. The idea is that you use a single algorithm for verification purposes. If you need to support different signing keys you can already use the JWKs by providing a public key in the well-known/jwks file and a key id claim in the tokens. By using the kid and the domain, that file is obtained along with the correct public key.

**Comment by jeacott1** [comment]
nice! not open for discussion in an open source project, awesome attitude.
how should your flow work for integration tests? I want to be able to have my tests generate tokens with a range of attributes/expiry/users etc to test my code, and I want to support production RSA tokens without resorting to fundamentally different code paths and configuration. the addition of a secret key used during testing and the change I suggest provides this perfectly.

**Comment by jeacott1** [comment]
Agreed the alg wont change during runtime, but I want to test my code using  various tokens in various states with varying content preferably using the same configuration, just with different parameters provided. in my proposed case I simply wouldn't provide a secret for production use.

As such my tests generate their own HMAC tokens - not dependent on auth0  - with which to properly test and get proper coverage. I cannot generate non auth0 dependent RSA tokens dynamically.
 
I am currently forced to provide multiple alternate configs using multiple conditional 
@Profile("test") @Profile("!test") type switches which as you point out really sux.

I'm not especially interested in testing your code, your tests already do that. if the spring integration works properly for 1 token alg  I'm confident it will work for others - its what happens after your code I particularly care about and need to test. 

I'd like to understand how using auth0 I should dynamically modify token expiry and content for multiple tests running in parallel? Or better without any auth0 service dependency at all?

**Comment by jeacott1** [comment]
Re "There's no "alg" value to switch on." 
- there is, although its not available via the interface it could easily be exposed.  all implementations carry the data, and as a hack is available brute forcing it something like:

((JWTDecoder)((PreAuthenticatedAuthenticationJsonWebToken) authentication).getDetails()).getAlgorithm()
or ((JWTDecoder)((AuthenticationJsonWebToken) authentication).getDetails()).getAlgorithm()
